### PR TITLE
add snippet.nvim support

### PR DESCRIPTION
**Pull request by salkin-mada**
1. 
Did use extra safety in `SCNvim.sc` -> `snippetFormat = snippetFormat ? "ultisnips";`
But really `util.vim` keeps check of this.

2.
Will add info about `vim.g.scnvim_snippet_format` in the wiki after merge.
Additionally mention the `vim.g.scdoc_render_prg` - i don't see it mentioned anywhere, but maybe i am mistaking.

PS. Not really sure about line `159` in `SCNvim.sc`, old line?

### CODE REVIEW

**Comment by madskjeldgaard** [comment]
Amazing work. I can test this out

**Comment by davidgranstrom** [comment]
@salkin-mada Thanks for this! I'll try and review this tonight.

> Did use extra safety in `SCNvim.sc` -> `snippetFormat = snippetFormat ? "ultisnips";`
> But really `util.vim` keeps check of this.
> 
> 
> Will add info about `vim.g.scnvim_snippet_format` in the wiki after merge.
> Additionally mention the `vim.g.scdoc_render_prg` - i don't see it mentioned anywhere, but maybe i am mistaking.

Please add a note about the new option in `doc/SCNvim.txt` with a description on what it does. I would maybe put it under the "Configuration" section among the "Extra" options. Also great if you could add it to the wiki of course!

> 
> PS. Not really sure about line `159` in `SCNvim.sc`, old line?

Can you send a link to the line in question here on github? I'm not sure either :)

**Comment by madskjeldgaard** [comment]
This works really wonderfully!

I wonder if we could make scnvim add the path of the resulting snippets to the package path automatically as part of this? So that the user only needs to `require'scnvim_snippets'` to get the table.

**Comment by davidgranstrom** [comment]
> This works really wonderfully!
> 
> I wonder if we could make scnvim add the path of the resulting snippets to the package path automatically as part of this? So that the user only needs to `require'scnvim_snippets'` to get the table.

Maybe we could add a function in `lua/scnvim/utils.lua` that reads the generated file and returns the table for convenience?

**Comment by salkin-mada** [comment]
wiki [PR](https://github.com/davidgranstrom/scnvim-wiki/pull/1)

**Comment by salkin-mada** [comment]
@madskjeldgaard @davidgranstrom nice idea.
would be great if the `scnvim_snippets.lua` would be automatically added to package.path.
will look into that.

**Comment by davidgranstrom** [comment]
> @madskjeldgaard @davidgranstrom nice idea.
> would be great if the `scnvim_snippets.lua` would be automatically added to package.path.
> will look into that.

I don't think we should modify package.path unless necessary, but to add a function that returns the (snippet) table would probably be useful! I'm not a snippets.nvim user so maybe I don't fully understand all the details, how does the configuration work?

**Comment by madskjeldgaard** [comment]
> @madskjeldgaard @davidgranstrom nice idea.
> would be great if the `scnvim_snippets.lua` would be automatically added to package.path.
> will look into that.

I tried messing around with it here and there is some complexity in it. If you are using scnvims root dir global var the package needs to be added after that var is populated. And then additionally the user can only require this module with the snippets AFTER scnvim has loaded. Hmm... maybe there should simply be an scnvim function one can call whenever and then that adds it to the package path and returns the module ? Just ideas....amaybe there's a cooler way of doing it

**Comment by salkin-mada** [comment]
> Can you send a link to the line in question here on github? I'm not sure either :)
in PR feature branch:
https://github.com/salkin-mada/scnvim/blob/assets/scide_scnvim/Classes/SCNvim.sc#L159

and in the master here:
https://github.com/davidgranstrom/scnvim/blob/master/scide_scnvim/Classes/SCNvim.sc#L153

wouldn't it be cleaner to say:
```supercollider
// vim
snippet = "snippet %\n%\nendsnippet\n".format(snippetName, className ++ argList);
// lua
snippet = "['%'] = [[%]];\n".format(snippetName, className ++ argList);
```
or actually just create a new var. `snippetContent`
```
snippetContent = className ++ argList;
                        case
                        {snippetFormat == "ultisnips"} {
                            snippet = "snippet %\n%\nendsnippet\n".format(snippetName, snippetContent);
                        }
                        {snippetFormat == "snippets.nvim"} {
                            snippet = "['%'] = [[%]];\n".format(snippetName, snippetContent);
                        };
```

But again. This is silly details :) - i just stumbled upon that line a bit.
Now i see that's it a simple redefinition of the `snippet` var. No problem really.

**Comment by salkin-mada** [comment]
> I don't think we should modify package.path unless necessary, but to add a function that returns the (snippet) table would probably be useful! I'm not a snippets.nvim user so maybe I don't fully understand all the details, how does the configuration work?

good point. I need to go AFK. Will return later with snippets.nvim config workings info. :)
But yes. `snippets.supercollider = X` really just need a lua table.

**Comment by madskjeldgaard** [comment]
> > I don't think we should modify package.path unless necessary, but to add a function that returns the (snippet) table would probably be useful! I'm not a snippets.nvim user so maybe I don't fully understand all the details, how does the configuration work?
> 
> good point. I need to go AFK. Will return later with snippets.nvim config workings info. :)
> But yes. `snippets.supercollider = X` really just need a lua table.

good ideas!

**Review by davidgranstrom — CHANGES_REQUESTED** [review]

File: scide_scnvim/Classes/SCNvim.sc [hunk 1]
```diff
@@ -151,16 +157,33 @@ SCNvim {
...
+        case
+        {snippetFormat == "ultisnips"} {
+            file.write("# SuperCollider snippets" ++ Char.nl);
+            file.write("# Snippet generator: SCNvim.sc" ++ Char.nl);
+            file.putAll(snippets);
+        }
+        {snippetFormat == "snippets.nvim"} {
+            file.write("-- SuperCollider snippets" ++ Char.nl);
+            file.write("-- Snippet generator: SCNvim.sc" ++ Char.nl);
+            file.write("M = {" ++ Char.nl);
```
[the remark this answers is not in the document]
Comment (salkin-mada) [reply scide_scnvim/Classes/SCNvim.sc:182]: true. will go for a lowercase local then -> `local snippets = ...` and `return snippets`

File: scide_scnvim/Classes/SCNvim.sc [hunk 2]
```diff
@@ -45,14 +45,19 @@ SCNvim {
         SkipJack(stlFunc, interval, name: "scnvim_statusline");
     }
 
-    *generateAssets {|rootDir|
+    *generateAssets {|rootDir, snippetFormat|
```
[the remark this answers is not in the document]
Comment (salkin-mada) [reply scide_scnvim/Classes/SCNvim.sc:48]: yep, kool. will add the changes now

File: autoload/scnvim/util.vim [hunk 3]
```diff
@@ -95,12 +95,29 @@ function! scnvim#util#get_scdoc_render_args() abort
   return get(g:, 'scnvim_scdoc_render_args', '% --from html --to plain -o %')
 endfunction
 
+function! scnvim#util#get_snippet_format() abort
```
[the remark this answers is not in the document]
Comment (salkin-mada) [reply autoload/scnvim/util.vim:98]: sure. nice!

**Review by salkin-mada — COMMENTED** [review]
pushed changes.

**Review by davidgranstrom — COMMENTED** [review]
@salkin-mada Thanks, looks good!

**Review by davidgranstrom — APPROVED** [review]

**Comment by madskjeldgaard** [comment]
If anyone was wondering how to use this PR with the lua-supercollider-snippets plugin, here is a temporary hack that I am sharing here in case anyone needs it :) 

```lua

local supercollider_snippets = require'supercollider-snippets'

local function append_scnvim_snippets(intable)
	local scnvim_snippet_path = "/home/mads/.local/share/nvim/plugged/scnvim/scnvim-data/scnvim_snippets.lua"
	local scnvimsnips = loadfile(scnvim_snippet_path)()
	for k,v in pairs(scnvimsnips) do
		intable[k] = v
	end

	return intable
end

append_scnvim_snippets(supercollider_snippets)

```

and then you can insert `supercollider_snippets` in the global snippets table

**Comment by davidgranstrom** [comment]
@salkin-mada I wrote a small lua helper that returns the current snippet file (based on the format). I guess it is mostly useful for this particular case, because it will make it easy to assign the snippets for `snippets.nvim` to use, but I made it general enough so that it also possible to read the contents of supercollider.snippets and could be extended to more formats later on. Is it okay to push it to this branch or do you have something locally that you would like to push first?

**Comment by salkin-mada** [comment]
@davidgranstrom sorry for the wait.
sounds good. push it.

**Comment by davidgranstrom** [comment]
> @davidgranstrom sorry for the wait.
> sounds good. push it.

No worries. Pushed. Let me know if you think this will work for the `snippets.nvim` use case and I'll include it!

**Comment by davidgranstrom** [comment]
> kool, looks fine.
> but could you show me how to use this function outside the scnvim dir.
> If I understand correctly this helper function returns either the string for ultisnips or the table for snippets.nvim? tried adding it to a util func `scnvim#utils#get_snippets` so i could call it via nvim call or vim.fn. But cant get it to return the table. please remind me / enlighten me on how this could be used. thanks

Sure! I was thinking something this might work (haven't had a chance to try `snippets.nvim` yet though..)

```lua
local scnvim_snippets = require'scnvim.util'.get_snippets()

if scnvim_snippets then
  -- snippets is pointing to snippets.nvim module
  snippets.supercollider = scnvim_snippets
end
```

**Comment by salkin-mada** [comment]
in `lua` i would do something like
```lua
local scnvim_snippets = require'scnvim.util'.get_snippets()
for k,v in pairs(scnvim_snippets) do snippets.supercollider[k] = v end
```
but i cant get `scnvim.util` into my path it seems. my package path loader cant find the module.
first i thought i forgot to pull your commits :) but that wasnt it.

cant get it to work like i would with:
```lua
-- snippets.supercollider already holds my own snippets
package.path = package.path .. assert(";/home/salkin/kode/scnvim-fork/scnvim/scnvim-data/scnvim_snippets.lua")
for k,v in pairs(require'scnvim_snippets') do snippets.supercollider[k] = v end
```
or
```lua
local supercollider_snippets = require'my-own-supercollider-snippets'
local scnvim_snippets = loadfile(";/home/salkin/kode/scnvim-fork/scnvim/scnvim-data/scnvim_snippets.lua")()
local function add_scnvim_snippets(table)
	for k,v in pairs(scnvim_snippets) do
		table[k] = v
	end
	return table
end
	
add_scnvim_snippets(supercollider_snippets)
```

**Comment by salkin-mada** [comment]
>  @madskjeldgaard
> 
> ... messing around with it here and there is some complexity in it. If you are using scnvims root dir global var the package needs to be added after that var is populated. And then additionally the user can only require this module with the snippets AFTER scnvim has loaded.  ...

My `invalid module`-trouble is probably a result of some init time thingy . Like the observations Mads was talking about.

**Comment by davidgranstrom** [comment]
> > @madskjeldgaard
> > ... messing around with it here and there is some complexity in it. If you are using scnvims root dir global var the package needs to be added after that var is populated. And then additionally the user can only require this module with the snippets AFTER scnvim has loaded.  ...
> 
> My `invalid module`-trouble is probably a result of some init time thingy . Like the observations Mads was talking about.

Hmm, yes that seems to be the case. Good call @madskjeldgaard! I think the solution is to move the `get_scnvim_root_dir` function from installation to the other utilities so we can call that function instead. I'll let you know as soon as I've tried it out locally.

**Comment by davidgranstrom** [comment]
@salkin-mada I moved the get_root_dir function to the other utilities, I think it should work as expected now.

**Comment by salkin-mada** [comment]
@madskjeldgaard could you test this snippets.nvim convenience table getter? You are on `vim-plug` if i remember correctly.
I am starting to suspect my lua package/module init time issues is due to some `packer.nvim` thingy..
possibly a direct consequence is something as puzzling as this: #140 
see [comment](https://github.com/davidgranstrom/scnvim/pull/140#issuecomment-812844273)

**Comment by madskjeldgaard** [comment]
I just tested the latest commits here and it seems to work nicely. I am now using the get_scnvim_root_dir() method to get the data and it works well ! 

Here's how I do it now for reference

```lua

local supercollider_snippets = require'supercollider-snippets'

local function append_scnvim_snippets(intable)
	local scnvim_snippet_path = require'scnvim/utils'.get_scnvim_root_dir() .. "/scnvim-data/scnvim_snippets.lua"
	local scnvimsnips = loadfile(scnvim_snippet_path)()
	for k,v in pairs(scnvimsnips) do
		intable[k] = v
	end

	return intable
end

append_scnvim_snippets(supercollider_snippets)
```

and then I put the `supercollider_snippets` variable in the snippets.nvim globa snippets l table.

**Comment by davidgranstrom** [comment]
> I just tested the latest commits here and it seems to work nicely. I am now using the get_scnvim_root_dir() method to get the data and it works well !
> 
> Here's how I do it now for reference
> 
> ```lua
> local supercollider_snippets = require'supercollider-snippets'
> 
> local function append_scnvim_snippets(intable)
> 	local scnvim_snippet_path = require'scnvim/utils'.get_scnvim_root_dir() .. "/scnvim-data/scnvim_snippets.lua"
> 	local scnvimsnips = loadfile(scnvim_snippet_path)()
> 	for k,v in pairs(scnvimsnips) do
> 		intable[k] = v
> 	end
> 
> 	return intable
> end
> 
> append_scnvim_snippets(supercollider_snippets)
> ```
> 
> and then I put the `supercollider_snippets` variable in the snippets.nvim globa snippets l table.

@madskjeldgaard Thanks for the feedback! There is also a convenience function added for this purpose called `get_snippets()` (also in utils). I image something like this should work (your example but modified)

```lua
local function append_scnvim_snippets(intable)
  local scnvimsnips = require'scnvim/utils'.get_snippets()
  for k,v in pairs(scnvimsnips) do
    intable[k] = v
  end
  return intable
end
```

Does it still work if you exchange your function for this one?

**Comment by madskjeldgaard** [comment]
> > I just tested the latest commits here and it seems to work nicely. I am now using the get_scnvim_root_dir() method to get the data and it works well !
> > Here's how I do it now for reference
> > ```lua
> > local supercollider_snippets = require'supercollider-snippets'
> > 
> > local function append_scnvim_snippets(intable)
> > 	local scnvim_snippet_path = require'scnvim/utils'.get_scnvim_root_dir() .. "/scnvim-data/scnvim_snippets.lua"
> > 	local scnvimsnips = loadfile(scnvim_snippet_path)()
> > 	for k,v in pairs(scnvimsnips) do
> > 		intable[k] = v
> > 	end
> > 
> > 	return intable
> > end
> > 
> > append_scnvim_snippets(supercollider_snippets)
> > ```
> > 
> > 
> > and then I put the `supercollider_snippets` variable in the snippets.nvim globa snippets l table.
> 
> @madskjeldgaard Thanks for the feedback! There is also a convenience function added for this purpose called `get_snippets()` (also in utils). I image something like this should work (your example but modified)
> 
> ```lua
> local function append_scnvim_snippets(intable)
>   local scnvimsnips = require'scnvim/utils'.get_snippets()
>   for k,v in pairs(scnvimsnips) do
>     intable[k] = v
>   end
>   return intable
> end
> ```
> 
> Does it still work if you exchange your function for this one?

Yep this works nicely! Thanks! IMO this PR is good now :)

**Comment by davidgranstrom** [comment]
@salkin-mada Thanks! Merging this now.

**Comment by salkin-mada** [comment]
🎉

**Pull request merged**